### PR TITLE
Implement PluginLoader

### PR DESCRIPTION
**Pull request by Mmduh-483**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -89,18 +89,37 @@ test-verbose: ARGS=-v            ## Run tests in verbose mode with coverage repo
 test-race:    ARGS=-race         ## Run tests with race detector
 $(TEST_TARGETS): NAME=$(MAKECMDGOALS:test-%=%)
 $(TEST_TARGETS): test
-check test tests: fmt lint | $(BASE) ; $(info  running $(NAME:%=% )tests...) @ ## Run tests
+pre-test-plugins: VALID_PLUGIN = $(CURDIR)/test/plugins/valid-plugin
+                  NO_PLUGIN_SYMBOL = $(CURDIR)/test/plugins/no-plugin-symbol
+                  NOT_VALID_PLUGIN_SYMBOL = $(CURDIR)/test/plugins/not-valid-plugin-symbol
+pre-test-plugins: | $(BASE) ; $(info Buidling test go plugins)
```
Comment (moshe010) [inline Makefile]: is this taken from SR-IOV operator

File: pkg/sm/plugin.go [hunk 2]
```diff
@@ -12,6 +19,28 @@ func NewPluginLoader() PluginLoader {
...
+	}
+
+	symbol, err := smPlugin.Lookup("Plugin")
+	if err != nil {
+		err = fmt.Errorf(`LoadPlugin(): failed to find "Plugin" object in the plugin file: %v`, err)
+		glog.Error(err)
+		return nil, err
+	}
+
+	smClient, ok := symbol.(plugins.SubnetManagerClient)
```
Comment (moshe010) [inline pkg/sm/plugin.go]: are you passing here mock? why?

File: Makefile (hunk 1 above)
Comment (Mmduh-483) [reply Makefile]: no

File: pkg/sm/plugin.go (hunk 2 above)
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: the `lookup` function returns a `Symbol`,  line 38 check if it is possible to cast Symbol to interface `SubnetManagerClient`, it is the same as the SR-IOV operator

File: Makefile (hunk 1 above)
Comment (moshe010) [reply Makefile]: so why do we need the pre-test-plugins? and why you have it twice

Comment (Mmduh-483) [reply Makefile]: for unit testing of the PluginLoader, it needs to load go lang plugins files `plugin.so` which is binary files needed to be build before running the unit test.

The first one define the Constants needed for the rule and the second is the work flow, it is the same as the `test-coverage`.

File: Makefile [hunk 3]
```diff
@@ -89,18 +89,37 @@ test-verbose: ARGS=-v            ## Run tests in verbose mode with coverage repo
...
+	@echo 'package main;var Plugin bool' > $(NOT_VALID_PLUGIN_SYMBOL)/main.go
+	@echo 'package main;import "github.com/Mellanox/ib-kubernetes/pkg/sm/plugins/mocks";var Plugin mocks.SubnetManagerClient;func init(){Plugin=mocks.SubnetManagerClient{}}' \
+	> $(VALID_PLUGIN)/main.go
+
+	$Q for plugin in $(VALID_PLUGIN) $(NO_PLUGIN_SYMBOL) $(NOT_VALID_PLUGIN_SYMBOL); do cd $$plugin && $(GO) build -buildmode=plugin main.go; done
+
+clean-pre-test-plugins: | pre-test-plugins ; $(info Remove test go plugins)
+	@rm -rf $(CURDIR)/test/plugins
+
+check test tests: fmt lint pre-test-plugins | $(BASE) ; $(info  running $(NAME:%=% )tests...) @ ## Run tests
```
Comment (moshe010) [inline Makefile]: remove fmt and lint

File: Makefile [hunk 4]
```diff
@@ -89,18 +89,37 @@ test-verbose: ARGS=-v            ## Run tests in verbose mode with coverage repo
...
+pre-test-plugins: VALID_PLUGIN = $(CURDIR)/test/plugins/valid-plugin
+                  NO_PLUGIN_SYMBOL = $(CURDIR)/test/plugins/no-plugin-symbol
+                  NOT_VALID_PLUGIN_SYMBOL = $(CURDIR)/test/plugins/not-valid-plugin-symbol
+pre-test-plugins: | $(BASE) ; $(info Buidling test go plugins)
+	$Q mkdir -p $(VALID_PLUGIN)
+	$Q mkdir -p $(NO_PLUGIN_SYMBOL)
+	$Q mkdir -p $(NOT_VALID_PLUGIN_SYMBOL)
+	@echo 'package main' > $(NO_PLUGIN_SYMBOL)/main.go
+	@echo 'package main;var Plugin bool' > $(NOT_VALID_PLUGIN_SYMBOL)/main.go
+	@echo 'package main;import "github.com/Mellanox/ib-kubernetes/pkg/sm/plugins/mocks";var Plugin mocks.SubnetManagerClient;func init(){Plugin=mocks.SubnetManagerClient{}}' \
```
Comment (moshe010) [inline Makefile]: create noop  plugin which just logs the action to syslog

File: Makefile [hunk 5]
```diff
@@ -89,18 +89,37 @@ test-verbose: ARGS=-v            ## Run tests in verbose mode with coverage repo
 test-race:    ARGS=-race         ## Run tests with race detector
 $(TEST_TARGETS): NAME=$(MAKECMDGOALS:test-%=%)
 $(TEST_TARGETS): test
-check test tests: fmt lint | $(BASE) ; $(info  running $(NAME:%=% )tests...) @ ## Run tests
+pre-test-plugins: VALID_PLUGIN = $(CURDIR)/test/plugins/valid-plugin
+                  NO_PLUGIN_SYMBOL = $(CURDIR)/test/plugins/no-plugin-symbol
```
Comment (moshe010) [inline Makefile]: we should have test symbole that will execute all unit test of all plugins

File: Makefile (hunk 3 above)
Comment (Mmduh-483) [reply Makefile]: removed

File: Makefile (hunk 4 above)
Comment (Mmduh-483) [reply Makefile]: added

File: Makefile (hunk 5 above)
Comment (Mmduh-483) [reply Makefile]: added rule for plugins

**Comment by moshe010** [comment]
can you increase coverage here?

**Comment by Mmduh-483** [comment]
> can you increase coverage here?

It is because of the no op plugin, it will be removed when adding the ufm plugin.
The code in the sm/plugins.go is covered by 100%

File: pkg/sm/plugins/plugin/no_op_plugin.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package main
```
Comment (moshe010) [inline pkg/sm/plugins/plugin/no_op_plugin.go]: rename file noop_plugin

File: pkg/sm/plugins/plugin/no_op_plugin.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+import "log"
+
+var Plugin plugin
+
+func init() {
+	Plugin = plugin{}
+	log.Println("No Op Plugin:", Plugin)
+}
```
Comment (moshe010) [inline pkg/sm/plugins/plugin/no_op_plugin.go]: s/No Op/noop in all sections

File: pkg/sm/plugins/plugin/no_op_plugin.go [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+type plugin struct {
+}
+
+func (p *plugin) Name() string {
+	log.Println("No Op Plugin Name()")
+	return ""
```
Comment (moshe010) [inline pkg/sm/plugins/plugin/no_op_plugin.go]: you should return "noop"

File: pkg/sm/plugins/plugin/no_op_plugin.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+func init() {
+	Plugin = plugin{}
+	log.Println("No Op Plugin:", Plugin)
+}
+
+type plugin struct {
+}
+
+func (p *plugin) Name() string {
+	log.Println("No Op Plugin Name()")
```
Comment (moshe010) [inline pkg/sm/plugins/plugin/no_op_plugin.go]: you don't need log message here

Comment (Mmduh-483) [reply pkg/sm/plugins/plugin/no_op_plugin.go]: fixed

File: pkg/sm/plugins/plugin/no_op_plugin.go (hunk 8 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin/no_op_plugin.go]: fixed

File: pkg/sm/plugins/plugin/no_op_plugin.go (hunk 7 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin/no_op_plugin.go]: fixed

File: pkg/sm/plugins/plugin/no_op_plugin.go (hunk 6 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin/no_op_plugin.go]: renamed

**Comment by Mmduh-483** [comment]
> covered

added

**Comment by Mmduh-483** [comment]
> you can mock the noop to verify that error cases are covered

added

File: pkg/sm/plugins/plugin/noop_plugin.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
+package main
+
+import "log"
```
Comment (adrianchiris) [inline pkg/sm/plugins/plugin/noop_plugin.go]: use glog ?

File: pkg/sm/plugin.go [hunk 11]
```diff
@@ -1,8 +1,19 @@
 package sm
 
+import (
+	"fmt"
+	"plugin"
+
+	"github.com/Mellanox/ib-kubernetes/pkg/sm/plugins"
+
+	"github.com/golang/glog"
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: It should be builtin, project pkgs, 3rd part
please check here https://github.com/golang/go/wiki/CodeReviewComments#imports

File: pkg/sm/plugin.go [hunk 12]
```diff
@@ -1,8 +1,19 @@
...
+)
+
+const DefaultPluginSymbolName = "Plugin"
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: It makes more flexibility for the variable name, and helps in unit test cases with the noop plugin to support negative cases.

File: pkg/sm/plugin.go [hunk 13]
```diff
@@ -1,8 +1,19 @@
...
+
 type PluginLoader interface {
-	// LoadPlugin load go plugin from given path, and return error if failed.
-	LoadPlugin(path string) error
+	// LoadPlugin loads go plugin from given path, it returns error if failed.
+	LoadPlugin(path, symbolName string) (plugins.SubnetManagerClient, error)
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: added

File: pkg/sm/plugin.go [hunk 14]
```diff
@@ -12,6 +23,28 @@ func NewPluginLoader() PluginLoader {
...
+	smPlugin, err := plugin.Open(path)
+	if err != nil {
+		err = fmt.Errorf("LoadPlugin(): failed to load plugin: %v", err)
+		glog.Error(err)
+		return nil, err
+	}
+
+	symbol, err := smPlugin.Lookup(symbolName)
+	if err != nil {
+		err = fmt.Errorf(`LoadPlugin(): failed to find "%s" object in the plugin file: %v`, symbolName, err)
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: fixed

File: pkg/sm/plugin.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (moshe010) [reply pkg/sm/plugin.go]: @Mmduh-483 the glog is builtin also it better to create PR which automatically check import order.

File: pkg/sm/plugins/plugin/noop_plugin.go (hunk 10 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/plugin/noop_plugin.go]: added

File: pkg/sm/plugin.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/sm/plugin.go]: so Define in the unit test :)

File: pkg/sm/plugin.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (adrianchiris) [reply pkg/sm/plugin.go]: @Mmduh-483 i see, thx, so ignore similar comments

[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: @moshe010  I will add one

File: pkg/sm/plugin.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/sm/plugin.go]: I will use it in the daemon also.
